### PR TITLE
ENH: pivot/groupby index with nan (GH3729)

### LINKED ISSUE #3729: ENH: pivot/groupby index with nan
ENH: maybe for now just provide a warning if dropping the nan rows when pivotting...

rom ml

http://stackoverflow.com/questions/16860172/python-pandas-pivot-table-silently-drops-indices-with-nans

This is effectivly trying to groupby on a NaN, currently not allowed

```
In [13]: a = [['a', 'b', 12, 12, 12], ['a', nan, 12.3, 233., 12], ['b', 'a', 123.23, 123, 1], ['a', 'b', 1, 1, 1.]]

In [14]: df = DataFrame(a, columns=['a', 'b', 'c', 'd', 'e'])

In [15]: df.groupby(['a','b']).sum()
Out[15]: 
          c    d   e
a b                 
a b   13.00   13  13
b a  123.23  123   1
```

Workaround to fill the index with a dummy, pivot, and replace

```

    In [31]: df2 = df.copy()

    In [32]: df2['dummy'] = np.nan

    In [33]: df2['b'] = df2['b'].fillna('dummy')

    In [34]: df2
    Out[34]: 
       a      b       c    d   e  dummy
    0  a      b   12.00   12  12    NaN
    1  a  dummy   12.30  233  12    NaN
    2  b      a  123.23  123   1    NaN
    3  a      b    1.00    1   1    NaN

    In [35]: df2.pivot_table(rows=['a', 'b'], values=['c', 'd', 'e'], aggfunc=sum)
    Out[35]: 
       a      b       c    d   e
    0  a      b   13.00   13  13
    1  a  dummy   12.30  233  12
    2  b      a  123.23  123   1

    In [36]: df2.pivot_table(rows=['a', 'b'], values=['c', 'd', 'e'], aggfunc=sum).replace('dummy',np.nan)
    Out[36]: 
       a    b       c    d   e
    0  a    b   13.00   13  13
    1  a  NaN   12.30  233  12
    2  b    a  123.23  123   1

```

### PR DESCRIPTION
**Pull request by andrewdalecramer**
Adds the ability to use NaN as a grouping variable. Required allowing NA as a factorisation variable in `factorize()`.

- [x] closes #3729
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/groupby/test_groupby.py [hunk 1]
```diff
@@ -276,6 +276,12 @@ def test_with_na_groups(dtype):
 
     assert_series_equal(agged, expected, check_dtype=False)
 
+    grouped_na = values.groupby(labels, group_nas=True)
+    agged_na = grouped_na.agg(len)
+    expected_na = Series([4, 2, 4], index=['bar', 'foo', np.nan])
+
+    assert_series_equal(agged_na, expected_na, check_dtype=False)
+
```
Comment (gfyoung) [inline pandas/tests/groupby/test_groupby.py]: Reference issue number in a comment.

File: pandas/tests/groupby/test_groupby.py [hunk 2]
```diff
@@ -288,6 +294,28 @@ def f(x):
...
+    )
+    assert_series_equal(agged, expected, check_dtype=False)
+
+    agged_na = df.groupby("Labs", group_nas=True)["Vals"].sum()
+    expected_na = Series(
+        [4, 2, 4], index=Index(['bar', 'foo', np.nan], name='Labs'),
+        name='Vals'
+    )
+    assert_series_equal(agged_na, expected_na, check_dtype=False)
+
```
Comment (gfyoung) [inline pandas/tests/groupby/test_groupby.py]: Reference issue number in a comment.

File: pandas/core/algorithms.py [hunk 3]
```diff
@@ -642,6 +645,13 @@ def factorize(values, sort=False, order=None, na_sentinel=-1, size_hint=None):
                                         na_sentinel=na_sentinel,
                                         assume_unique=True)
 
+    if factor_nas and (labels == na_sentinel).any():
+        new_na_sentinel = len(uniques)
+        # Append the np.nan
+        uniques = np.resize(uniques, len(uniques) + 1)
+        uniques[-1] = np.nan
+        labels = np.where(labels == na_sentinel, new_na_sentinel, labels)
+
```
Comment (gfyoung) [inline pandas/core/algorithms.py]: Given that you're using broadcasting and `NumPy`, I don't think we should have any performance issues, but I wonder if we should still add a benchmark test anyway.

**Review by jreback — CHANGES_REQUESTED** [review]
needs quite a bit of tests & to parameterize existing tests to accept nan groups. i don't like the keyword at all. haven't reviewed fully.

**Comment by jreback** [comment]
@andrewdalecramer 

this needs to be ``dropna=None`` as a default, with a warning if its *not* passed and a default of ``True`` for back-compat. Eventually we would want to make this False.

**Comment by ppallesen** [comment]
@andrewdalecramer 
Suggestion for keyword name: allow_na_groups = False

**Comment by jreback** [comment]
can you rebase this

**Comment by jreback** [comment]
this is a nice feature, but needs quite a bit of work to fully test this out.

closing as stale. if you'd like to continue, pls ping.

**Pull request closed without merging**